Assets/MRTK/Providers/LeapMotion/LeapMotionDeviceManager.cs

```diff
@@ -0,0 +1,258 @@
+
+                // Set the leap attachment hand to the corresponding handedness
+                if (handedness == Handedness.Left)
+                {
+                    leapHand.SetAttachmentHands(leftAttachmentHand, LeapMotionServiceProvider);
+                }
+                else // handedness == Handedness.Right
+                {
+                    leapHand.SetAttachmentHands(rightAttachmentHand, LeapMotionServiceProvider);
+                }
+                
+                // Set the pointers for an articulated hand to the leap hand
+                foreach (var pointer in pointers)
+                {
+                    pointer.Controller = leapHand;
+                }
+
+                trackedHands.Add(handedness, leapHand);
+
+                CoreServices.InputSystem.RaiseSourceDetected(inputSource, leapHand);
+            }
+        }
+
+        /// <summary>
+        /// Removes the LeapMotionArticulated hand from the scene when the tracking is lost.
+        /// </summary>
+        /// <param name="handedness">The handedness (Handedness.Left or Handedness.Right) of the hand to be removed</param>
+        private void OnHandDetectionLost(Handedness handedness)
+        {
+            if (CoreServices.InputSystem != null)
+            {
+                CoreServices.InputSystem.RaiseSourceLost(trackedHands[handedness].InputSource);
+            }
+
+            // Disable the pointers if the hand is not tracking
+            RecyclePointers(trackedHands[handedness].InputSource);
+
+            // Remove hand from tracked hands
+            trackedHands.Remove(trackedHands[handedness].ControllerHandedness);
+        }
+
+        /// <summary>
+        /// Update the number of tracked leap hands.
+        /// </summary>
+        /// <param name="isLeftTracked">The tracking state of the left leap hand</param>
+        /// <param name="isRightTracked">The tracking state of the right leap hand</param>
+        private void UpdateLeapTrackedHands(bool isLeftTracked, bool isRightTracked)
+        {
+            // Left Hand Update
+            if (isLeftTracked && !trackedHands.ContainsKey(Handedness.Left))
+            {
+                OnHandDetected(Handedness.Left);
+            }            
+            else if (!isLeftTracked && trackedHands.ContainsKey(Handedness.Left))
+            {
+                OnHandDetectionLost(Handedness.Left);
+            }
+
+            // Right Hand Update
+            if (isRightTracked && !trackedHands.ContainsKey(Handedness.Right))
+            {
+                OnHandDetected(Handedness.Right);
+            }
+            else if (!isRightTracked && trackedHands.ContainsKey(Handedness.Right))
+            {
+                OnHandDetectionLost(Handedness.Right);
+            }
+        }
+
+        /// <inheritdoc />
+        public override void Update()
+        {
+            using (UpdatePerfMarker.Auto())
+            {
+                base.Update();
```

Comment:
@davidkline-ms for perf markers in subclasses, do we recommend having the base function be invoked OUTSIDE of the subclass's marker?

i.e. should base.Update() be moved outside of this using clause? Otherwise this marker ends up capturing the work of the base class (which I believe has its own instrumentation).

Reply:
I'd vote for leaving it as-is. When it's in scope like this and the base class has its own marker, the Unity profiler shows these nested in the call hierarchy and it becomes more obvious which class' base is being called.